### PR TITLE
SONAR-8772 Performance hotspot in verification of project permissions

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: sonar-db/src/main/java/org/sonar/db/permission/AuthorizationDao.java [hunk 1]
```diff
@@ -117,6 +133,10 @@ public int countUsersWithGlobalPermissionExcludingUserPermission(DbSession dbSes
       });
   }
 
+  /**
+   * @deprecated it loads too many results and there's no functional need.
+   */
+  @Deprecated
   public Collection<String> selectAuthorizedRootProjectsUuids(DbSession dbSession, @Nullable Integer userId, String role) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/permission/AuthorizationDao.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java [hunk 2]
```diff
@@ -152,21 +149,29 @@ protected boolean hasOrganizationPermissionImpl(String organizationUuid, String
...
+    if (permissionsByProjectUuid == null) {
+      permissionsByProjectUuid = HashMultimap.create();
     }
-    return projectUuidByPermission.get(permission).contains(projectUuid);
+    Set<String> permissions;
+    if (permissionsByProjectUuid.containsKey(projectUuid)) {
+      permissions = permissionsByProjectUuid.get(projectUuid);
+    } else {
+      permissions = loadProjectPermissions(projectUuid);
+      permissionsByProjectUuid.putAll(projectUuid, permissions);
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: according to javadoc of `putAll`: 
> In particular, this is a no-op if {@code values} is empty.

I'm wondering there is actually any caching when there is no permission for the projectUuid.
If it's a no-op, then the key won't be added to the Multimap and `containsKey` will always return `false`, hence no caching the next time method is called with the same `projectUuid`.

In which case, a `Map<String, Set<String>>` should be used instead (and the call to `containsKey` dropped in favor of a single call to `get`)

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java [hunk 3]
```diff
@@ -152,21 +149,29 @@ protected boolean hasOrganizationPermissionImpl(String organizationUuid, String
...
-        Collection<String> projectUuids = dbClient.authorizationDao().selectAuthorizedRootProjectsUuids(dbSession, getUserId(), permission);
-        addProjectPermission(permission, projectUuids);
-      }
+    if (permissionsByProjectUuid == null) {
+      permissionsByProjectUuid = HashMultimap.create();
     }
-    return projectUuidByPermission.get(permission).contains(projectUuid);
+    Set<String> permissions;
+    if (permissionsByProjectUuid.containsKey(projectUuid)) {
+      permissions = permissionsByProjectUuid.get(projectUuid);
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: why not `return permissionsByProjectUuid.get(projectUuid).contains(permission)` and drop the `else` and the undefined variable?

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java [hunk 4]
```diff
@@ -152,21 +149,29 @@ protected boolean hasOrganizationPermissionImpl(String organizationUuid, String
...
   }
 
-  private void addProjectPermission(String permission, Collection<String> authorizedProjectUuids) {
-    for (String key : authorizedProjectUuids) {
-      projectUuidByPermission.put(permission, key);
+  private Set<String> loadProjectPermissions(String projectUuid) {
+    Set<String> permissions;
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      if (userDto != null && userDto.getId() != null) {
+        permissions = dbClient.authorizationDao().selectProjectPermissions(dbSession, projectUuid, userDto.getId());
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: I favor multiple returns over undefined variable at the beginning of method. So old-school procedural style... :p

File: server/sonar-server/src/test/java/org/sonar/server/user/ServerUserSessionTest.java [hunk 5]
```diff
@@ -80,6 +80,15 @@ public void setUp() throws Exception {
   }
 
   @Test
+  public void anonymous_is_not_logged_in_and_does_not_have_login() throws Exception {
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/user/ServerUserSessionTest.java]: I don't see any test of the caching mechanism, except for `hasSystemAdminister` and `getGroups`. From statically looking at the code, this is a miss (see my other comment).

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java (hunk 4 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: I'm surprised myself.

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java [hunk 6]
```diff
@@ -52,12 +49,11 @@
   private final DbClient dbClient;
   private final OrganizationFlags organizationFlags;
   private final DefaultOrganizationProvider defaultOrganizationProvider;
-  private final Supplier<List<GroupDto>> groups = Suppliers.memoize(this::loadGroups);
+  private final Supplier<Collection<GroupDto>> groups = Suppliers.memoize(this::loadGroups);
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: I think this should be a Set and equals/hashcode should be defined in GroupDto (based on the id)

It's important we know the groups are not ordered, but also as important that there is no duplicate.

WDYT?

Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: yes, good idea. Out of scope of this PR so I will address it just after.

**Pull request closed without merging**